src/Renderer/NavigatorRenderer.cpp

```diff
@@ -0,0 +1,378 @@
+  
+  auto current_speed_s = static_cast<std::string>(current_speed);
+    
+  TextRenderer text_renderer1;
+  text_renderer1.SetVCenter(true);
+  text_renderer1.SetCenter(false);
+  text_renderer1.SetControl(true);
+
+  TextRenderer text_renderer2;
+  text_renderer2.SetVCenter(true);
+  text_renderer2.SetCenter(true);
+  text_renderer2.SetControl(true);
+  
+  canvas.Select(*look.font);
+
+  bool has_started = calculated.ordered_task_stats.start.task_started;
+
+  const auto time_elapsed = TimeStamp{FloatDuration{calculated.ordered_task_stats.total.time_elapsed}};
+  const RoughTimeDelta r{};
+
+  auto time_elapsed_s = FormatLocalTimeHHMM(time_elapsed, r).c_str();
+  has_started ? time_elapsed_s : time_elapsed_s = "--:--";
+
+  const auto time_start = calculated.ordered_task_stats.start.time;
+  auto time_start_s = FormatLocalTimeHHMM(time_start,
+                                  CommonInterface::GetComputerSettings().utc_offset).c_str();
+  has_started ? time_start_s : time_start_s = "--:--";
+
+  const auto time_local = FormatLocalTimeHHMM(basic.time,
+                                  CommonInterface::GetComputerSettings().utc_offset).c_str();
+
+  const auto time_planned = TimeStamp{FloatDuration{calculated.ordered_task_stats.total.time_planned}};
+  auto time_planned_s = FormatLocalTimeHHMM(time_planned, r).c_str();
+  has_started ? time_planned_s : time_planned_s = "--:--";
+
+  const auto arrival_planned = TimeStamp{FloatDuration{time_start.ToDuration() + time_planned.ToDuration()}};
+  auto arrival_planned_s = FormatLocalTimeHHMM(arrival_planned, 
+                                  CommonInterface::GetComputerSettings().utc_offset).c_str(); 
+  has_started ? arrival_planned_s : arrival_planned_s = "--:--";
+
+  auto waypoint_s = static_cast<std::string>(wp_current.name);
+
+  // WP Dist
+  const auto waypoint_distance = calculated.ordered_task_stats.current_leg.vector_remaining.distance;
+
+  std::ostringstream oss;
+  oss << std::fixed << std::setprecision(2) << std::noshowpoint  << waypoint_distance/1000; 
+  auto waypoint_distance_s = oss.str(); 
+
+  // WP ALTD
+  const auto waypoint_altitude_diff = static_cast<int>(calculated.ordered_task_stats.current_leg.solution_remaining.GetRequiredAltitude());
+  auto waypoint_altitude_diff_s = std::to_string(waypoint_altitude_diff); 
+
+  // WP GR
+  const auto waypoint_GR = calculated.ordered_task_stats.current_leg.gradient;
+  auto waypoint_GR_s = std::to_string(static_cast<int>(waypoint_GR));
+
+  // e_WP_BearingDiff
+  Angle bearing_diff{};
+  bearing_diff.Zero();
+  if (basic.track_available) 
+    bearing_diff = calculated.ordered_task_stats.current_leg.vector_remaining.bearing - basic.track;
+  auto waypoint_direction = static_cast<int>(bearing_diff.AsDelta().Degrees());
+  auto waypoint_direction_s = std::to_string(waypoint_direction);
+
+  const auto caption1 = waypoint_distance_s + " km | " + waypoint_altitude_diff_s + " m | " + waypoint_GR_s + ":1";
```

Comment:
Please don't concatenate strings using `std::string`, it's uncontrollable bloat.

Reply:
ok, amended as follows:
```
  TCHAR informations_next_waypoint_s[100]{}; 
  _stprintf(informations_next_waypoint_s, _T("%s | %s | %s | %s | %s"), 
            waypoint_distance_s, waypoint_altitude_diff_s, waypoint_GR_s, current_speed_s, waypoint_direction_s);
```
is it OK?

Reply:
Can you guarantee that the buffer will not overflow?

Reply:
In this case, I made a mistake, I cannot guarantee that the buffer will not overflow.
my assumption: each char array have a size of 20 --> it means 19 characters x 5  + 4 x " | " (three characters) + +1 char '\0'
result: 108 --> overflow (!)

so I will change this with appropriate length --> in this case : informations_next_waypoint_s[110] ;

Reply:
You could just use a variant of sprintf which truncates instead of overflowing the buffer. That way, you don't need an unrealistic worst-case stack allocation.

Reply:
I found the StaticString<..> very useful and easier to read ; That's why I use it now for all strings in the method NavigatorRender::drawText() ; in the background, it is using the snprintf() function which was the one I would like to use as you suggested.

as instance:
```c++
StaticString<8> time_start_s;
const auto time_start = calculated.ordered_task_stats.start.time;
if (tp == TaskType::ORDERED && has_started && basic.time_available)
  time_start_s.Format("%s", FormatLocalTimeHHMM(time_start, utc_offset1).c_str());
else
  time_start_s.Format("%s", "--:--");
```
instead of:
```c++
TCHAR time_start_s[10];
const auto time_start = calculated.ordered_task_stats.start.time;
if (tp == TaskType::ORDERED && has_started && basic.time_available)
  snprintf(time_start_s, ARRAY_SIZE(time_start_s), _T("%s"),
           FormatLocalTimeHHMM(time_start, utc_offset1).c_str());
else
  snprintf(time_start_s, ARRAY_SIZE(time_start_s), _T("%s"), "--:--");
```

Is it OK for you ?